store/sqlstore/channel_store.go

```diff
@@ -3132,3 +3401,287 @@ func (s SqlChannelStore) GroupSyncedChannelCount() (int64, *model.AppError) {
+		}
+
+		// clean previous favorites
+		sql, args, _ := s.getQueryBuilder().Delete("SidebarChannels").Where(sq.Eq{"ChannelId": oldCategory.Channels}).ToSql()
+
+		if _, err = transaction.Exec(sql, args...); err != nil {
+			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		if category.Type == model.SidebarCategoryFavorites {
+			// clean previous preference favorites (to support bi-directional sync)
+			sql, args, _ = s.getQueryBuilder().Delete("Preferences").Where(sq.And{
+				sq.Eq{"ChannelId": oldCategory.Channels},
+				sq.Eq{"Preferences.Category": model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL},
+			}).ToSql()
+
+			if _, err = transaction.Exec(sql, args...); err != nil {
+				return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+		var channels []interface{}
+		var preferences []interface{}
+		runningOrder := 0
+		for _, channelID := range oldCategory.Channels {
+			channels = append(channels, &model.SidebarChannel{
+				ChannelId:  channelID,
+				CategoryId: category.Id,
+				SortOrder:  int64(runningOrder),
+				UserId:     userId,
+			})
+			preferences = append(preferences, &model.Preference{
+				Name:     channelID,
+				UserId:   userId,
+				Category: model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL,
+				Value:    "true",
+			})
+			runningOrder += model.MinimalSidebarSortDistance
+		}
+
+		if err = transaction.Insert(channels...); err != nil {
+			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+
+		}
+
+		if category.Type == model.SidebarCategoryFavorites {
+			if err = transaction.Insert(preferences...); err != nil {
+				return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+	}
+
+	if err = transaction.Commit(); err != nil {
+		return nil, model.NewAppError("SqlChannelStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.commit_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return categories, nil
+}
+
+// UpdateSidebarChannelByPreference is called when the Preference table is being updated to keep SidebarCategories in sync
+func (s SqlChannelStore) UpdateSidebarChannelByPreference(preference *model.Preference) *model.AppError {
+	if preference.Category != model.PREFERENCE_CATEGORY_DIRECT_CHANNEL_SHOW && preference.Category != model.PREFERENCE_CATEGORY_GROUP_CHANNEL_SHOW && preference.Category != model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL {
+		return nil
+	}
+	params := map[string]interface{}{
+		"UserId":       preference.UserId,
+		"ChannelId":    preference.Name,
```

Comment:
For DMs, the preference name isn't the channel ID of the DM unfortunately, it's the ID of the teammate that the DM is with.

Reply:
wat? :)
so there is not Channel with the Id of preference.Name for DMs? What about GMs?

Reply:
The preference's name is the channel ID for GMs, but it's the ID of the other user for DMs. If you have the ID of the current user, you can get the other user's ID  from the channel using [`Channel.GetOtherUserIdForDM`](https://github.com/mattermost/mattermost-server/blob/master/model/channel.go#L310).

I swear making it the ID of the other user made sense back when I added Preferences for showing/hiding DMs, but it definitely doesn't make sense looking back on it, especially since GMs were added

Reply:
ok, discussed this with @hmhealey offline, for now this function will only handle favorites